llvm/lib/SYCL/KernelProperties.cpp

```diff
@@ -10,72 +10,125 @@
+    for (Instruction &I : instructions(F)) {
+      auto *CB = dyn_cast<CallBase>(&I);
+      if (!CB || CB->getIntrinsicID() != Intrinsic::var_annotation)
+        continue;
+      auto *Alloca =
+          dyn_cast_or_null<AllocaInst>(getUnderlyingObject(CB->getOperand(0)));
+      // SYCL buffer's property for DDR bank association is lowered
+      // as an annotation. As the signature of
+      // llvm.var.annotate takes an i8* as first argument, cast from original
+      // argument type to i8* is done, and the final bitcast is annotated.
+      if (!Alloca)
+        continue;
+      auto *Str = cast<ConstantDataArray>(
+          cast<GlobalVariable>(getUnderlyingObject(CB->getOperand(1)))
+              ->getOperand(0));
+      if (Str->getRawDataValues() != kindOf("xilinx_ddr_bank"))
+        continue;
+      Constant *Args =
+          (cast<GlobalVariable>(getUnderlyingObject(CB->getOperand(4)))
+               ->getInitializer());
+      unsigned Bank;
+      if (auto *ZeroData = dyn_cast<ConstantAggregateZero>(Args))
+        Bank = 0;
+      else
+        Bank = cast<ConstantInt>(Args->getOperand(0))->getZExtValue();
 
-    userSpecifiedDDRBanks[Alloca] = Bank;
-    if (maxiBundles.find(Bank) == maxiBundles.end()) {
-      maxiBundles[Bank] = {formatv("ddrmem{0}", Bank)};
+      UserSpecifiedDDRBanks[Alloca] = Bank;
     }
   }
-}
 
-Optional<unsigned> KernelProperties::getUserSpecifiedDDRBank(Argument *Arg) {
+  Optional<unsigned> getUserSpecifiedDDRBank(Argument *Arg, SmallDenseMap<llvm::AllocaInst *, unsigned, 16> &UserSpecifiedDDRBanks) {
   for (User *U : Arg->users()) {
     if (auto *Store = dyn_cast<StoreInst>(U))
       if (Store->getValueOperand() == Arg) {
-        auto Lookup = userSpecifiedDDRBanks.find(dyn_cast_or_null<AllocaInst>(
+        auto Lookup = UserSpecifiedDDRBanks.find(dyn_cast_or_null<AllocaInst>(
             getUnderlyingObject(Store->getPointerOperand())));
-        if (Lookup == userSpecifiedDDRBanks.end())
+        if (Lookup == UserSpecifiedDDRBanks.end())
           continue;
         return {Lookup->second};
       }
   }
   return {};
 }
+} // namespace
 
-Optional<StringRef> KernelProperties::getArgumentMAXIBundle(Argument *Arg) {
-  auto DdrId = getUserSpecifiedDDRBank(Arg);
-  if (DdrId) {
-    return {maxiBundles[DdrId.getValue()].bundleName};
+namespace llvm {
+bool KernelProperties::isArgBuffer(Argument* Arg, bool SyclHLSFlow) {
+  if (Arg->getType()->isPointerTy() &&
+      (SyclHLSFlow ||
+       Arg->getType()->getPointerAddressSpace() == SPIRAS_Global ||
+       Arg->getType()->getPointerAddressSpace() == SPIRAS_Constant)) {
+    return !Arg->hasByValAttr();
   }
-  return {};
+  return false;
+}
+
+KernelProperties::KernelProperties(Function &F, bool SyclHlsFlow) {
+
+  SmallDenseMap<llvm::AllocaInst*, unsigned, 16> UserSpecifiedDDRBanks{};
+  collectUserSpecifiedDDRBanks(F, UserSpecifiedDDRBanks);
+
+  SmallDenseMap<Argument *, unsigned, 16> DDRAssignment{};
```

Comment:
By reading the uncommented code, it is not obvious what is happening, for example how are aggregated the DDR banks so there is only 1 bundle per unique DDR bank